### DISCUSSION TITLE
Export `WithImmer` type

### DISCUSSION THREAD

**Discussion by illarionvk** [pmndrs/zustand#1454] [Q&A]
## Summary

The VSCode editor reports that the inferred type is not portable and advises adding a type annotation. Typescript CLI (v4.8.4) doesn't report this as error.

The inferred type is:

```ts
UseBoundStore<WithImmer<StoreApi<AppState>>>
```

The `AppState` type is my custom type. I can import `UseBoundStore` and `StoreApi` from zustand, but `WithImmer` type is not exported from the immer middleware.

Could you export the `WithImmer` type as well?

## Link to reproduction

<img width="1208" alt="vscode-message" src="https://user-images.githubusercontent.com/5012724/204561357-a4f8735d-0764-47e4-8de5-5252707d6cab.png">

## Check List

Please do not ask questions in issues.

- [x] I understand this is not a question.

Please fill this template if you're filling an issue regarding TypeScript.

- [ ] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction. Or some other way of reproduction if these two don't suffice.
- [x] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by dai-shi** [comment]
It's actually intentional. We consider `WithImmer` is an internal type.
I might be missing something in the use case, though.
Can you please share typescript playground link?
Or, does it only happen locally with `--declaration` enabled?

**Reply by illarionvk** [reply]
I created a sandbox, but it doesn't surface the TS2742 error:

https://codesandbox.io/s/zustand-immer-ts2742-ybm9kn?file=/src/App.tsx

After reading the discussion in [Typescript issue 29808](https://github.com/microsoft/TypeScript/issues/29808#issuecomment-829750974), I resolved the error by directly importing the Draft type from the `immer` package in my file:

```
export type { Draft } from 'immer'
```
_I re-export the type to avoid 'unused type' ESLint warnings_

From my personal experience, I'd prefer all inferred types to be importable so I can explicitly declare what type I expect for my variable.

Could you please make `WithImmer` type public if it makes sense from the library standpoint?

<figure>
<figcaption><em>Middleware type declaration uses Draft type from immer</em></figcaption>
<img width="998" alt="" src="https://user-images.githubusercontent.com/5012724/204777231-15dfd33b-1008-4d2e-b384-be99549b0736.png">

</figure>

**Reply by dai-shi** [reply]
Please also see #930 discussion.
But is this specific to immer middleware which import a type from outside? 🤔

**Comment by pedroscosta** [comment]
Reopening this discussion, because it's still relevant.

Using the solution from #930 isn't ideal, because it requires every file that imports the mutated type to also import Immer and Devtools, otherwise, the type appear as _never_.

An example of this problem can be seen [here](https://github.com/pedroscosta/blue-fire/blob/4955a436a1fe1324e2abc55a4a3c0651390fad92/exports/bluefire.ts#L175).

Exporting the intermediary types such as _WithImmer_ and _WithDevtools_ fixes these problems all together and makes store typing a lot easier. So the code would look like this:
```typescript
export type BluefireStore = UseBoundStore<WithImmer<WithDevtools<StoreApi<BluefireState>>>>;
```
and the problem is fixed:
![image](https://user-images.githubusercontent.com/5331735/229908130-79c54d26-fdf5-4042-9ba2-8b4ec28f597d.png)

**Reply by dai-shi** [reply]
I don't know if we would like to change types on zustand end.
This case is kind of a library use case. So, the basic idea is to define types in your library.
I'm not trying to convince you. But, what I would do would be something like this:

```ts
type SetStateWithImmer<T> = (
  nextState: T | Partial<T> | ((state: Draft<T>) => void),
  replace?: boolean
) => void;

export type BluefireStore = UseBoundStore<
  Omit<StoreApi<BluefireState>, 'setState'> & {
    setState: SetStateWithImmer<BluefireState>
  }
>;
```
I'm not sure I want to expose devtools types in BluefireStore. (devtools shouldn't be enabled in production?)
WithImmer is over complicated for this case and we don't consider it a public api at the moment (can change in the future.) SetStateWithImmer could be exposed from zustand.

**Reply by pedroscosta** [reply]
In my personal opinion, exporting all types is a a good practice, except when they are strictly for internal use (which doesn't seem to be the case here for zustand), because it reduces boilerplate typing for people using the package. But I completely understand your point. 

In this particular case, I wouldn't export `SetStateWithImmer` from this package because whoever is using that shouldn't be creating a store, because this is an API for extensions, even if they did, they would have no need (or means) to export their store to any another extension.

Nonetheless, exporting some types like `SetStateWithImmer` from zustand is a good step in that direction and I would be really happy to see that going forward.

**Comment by alexandrutocar** [comment]
Hey, as some have already stated, exporting `WithImmer` would the optimal cover for TS2742, but also useful for some who write after explicit type declarations. We all could just rewrite `WithImmer` ourselves as an util-type, but to do it for every project, yet alone writing a dedicated util-type library to avoid re-exports, or forking the zustand repo and patching the types to be exported, seems as requiring too much hustle for me, and for others too. If there was an easier path, the issue would likely not have been raised at all. 😅

Besides that, the type does not seem internal, as it is something we, as the end users of the library have to deal with.

**Reply by dai-shi** [reply]
Let's wait for @devanshj thoughts on it.

**Comment by devanshj** [comment]
I'll let @dai-shi decide whether to export the internal types or not (I personally am inclined to not exporting too). I'm not sure how you would solve this problem, one solution that I think might work is something like this...

```ts
// @file zustand/middleware/immer.ts

import type { Draft as _Draft } from "immer"
type Draft<T> = _Draft<T>

// ...
```

Maybe someone can try it out and send a PR if it works.

**Comment by Cow258** [comment]
Here are the guide for use immer with zustand in JSDoc without using `WithImmer`

```js
/**
 * @typedef YourStoreTypes
 * @property {Object} foo
 * @property {string} foo.bar
 * @property {(bar: string) => void} updateBar
 */

export const useStore = (
  /** @type {typeof create<YourStoreTypes, [["zustand/immer", never]]>} */(create)(
  immer(
    (set) => ({
      foo: {
        bar: 'bar',
      },
      updateBar: (bar) => {
        set(state => {
          state.foo.bar = bar
        })
      },
    })
  )
))
```

![image](https://github.com/user-attachments/assets/7d1c57d9-bb19-42c1-96c9-e512d5d7fab0)


You can see more detail here
https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md
